tests/MemcacheTest.php

```diff
@@ -0,0 +1,140 @@
+    {
+        $count = 0;
+        $key = 'foo';
+        $value = 'bar';
+        $compute = function () use (&$count, $value) {
+            $count++;
+
+            return $value;
+        };
+
+        $client = $this->getMemcacheMock();
+        $client->expects(
+            $this->once()
+        )->method('get')->with($this->equalTo($key))->will($this->throwException(new \Exception()));
+
+        $memoizer = new Memcache($client);
+
+        $this->assertSame($value, $memoizer->memoizeCallable($key, $compute));
+        $this->assertSame(1, $count);
+    }
+
+    /**
+     * @test
+     * @covers ::__construct
+     * @covers ::memoizeCallable
+     */
+    public function memoizeCallableWithUncachedKey()
+    {
+        $count = 0;
+        $key = 'foo';
+        $value = 'bar';
+        $cachedValue = json_encode(['result' => $value]);
+        $cacheTime = 1234;
+        $compute = function () use (&$count, $value) {
+            $count++;
+
+            return $value;
+        };
+
+        $client = $this->getMemcacheMock();
+        $client->expects($this->once())->method('get')->with($this->equalTo($key))->will($this->returnValue(null));
+        $client->expects($this->once())->method('set')
+            ->with($this->equalTo($key), $this->equalTo($cachedValue), $this->equalTo(0), $this->equalTo($cacheTime));
+
+        $memoizer = new Memcache($client);
+
+        $this->assertSame($value, $memoizer->memoizeCallable($key, $compute, $cacheTime));
+        $this->assertSame(1, $count);
+    }
+
+    /**
+     * @test
+     * @covers ::__construct
+     * @covers ::memoizeCallable
+     */
+    public function memoizeCallableWithUncachedKeyWithExceptionOnSet()
+    {
+        $count = 0;
+        $key = 'foo';
+        $value = 'bar';
+        $cachedValue = json_encode(['result' => $value]);
+        $compute = function () use (&$count, $value) {
+            $count++;
+
+            return $value;
+        };
+
+        $client = $this->getMemcacheMock();
+        $client->expects(
+            $this->once()
+        )->method('get')->with($this->equalTo($key))->will($this->returnValue(null));
+        $setExpectation = $client->expects(
+            $this->once()
+        )->method('set')->with($this->equalTo($key), $this->equalTo($cachedValue));
+        $setExpectation->will($this->throwException(new \Exception()));
+
+        $memoizer = new Memcache($client);
+
+        $this->assertSame($value, $memoizer->memoizeCallable($key, $compute));
+        $this->assertSame(1, $count);
+    }
+
+    public function getMemcacheMock() : \Memcache
+    {
+        if (PHP_VERSION_ID < 80000) {
+            require_once 'MemcacheMockable.php';
+            return $this->getMockBuilder(MemcacheMockable::class)
+                ->setMethods(['get', 'set'])->getMock();
+        } else {
+            return $this->getMockBuilder(\Memcache::class)->setMethods(['get', 'set'])->getMock();
+        }
```

Comment:
```suggestion
        $isOlderPHPVersion = PHP_VERSION_ID < 80000;
        $memcacheClass =$isOlderPHPVersion ? MemcacheMockable::class : Memcache::class;
        return $this->getMockBuilder(\Memcache::class)->setMethods(['get', 'set'])->getMock();
```
Just a suggestion to avoid the inline `require_once` call